libdrgn/symbol.c

```diff
@@ -174,3 +178,286 @@ void drgn_symbol_result_builder_array(struct drgn_symbol_result_builder *builder
+		free(copy);
+		return false;
+	}
+	return true;
+}
+
+struct drgn_error *
+drgn_symbol_index_find(const char *name, uint64_t address,
+		       enum drgn_find_symbol_flags flags, void *arg,
+		       struct drgn_symbol_result_builder *builder)
+{
+	struct drgn_symbol_index *index = arg;
+
+	// Unlike the ELF symbol finder, we don't have any particular rules
+	// about which symbols get priority when looking up a single symbol.
+	// If we decide this logic is critical, it would probably make sense to
+	// move it into the symbol finder's API via the result builder, rather
+	// than reimplementing it here.
+
+	if (flags & DRGN_FIND_SYMBOL_ADDR) {
+		uint32_t start, end;
+		address_search_range(index, address, &start, &end);
+		for (uint32_t i = start; i < end; i++) {
+			struct drgn_symbol *s = &index->symbols[i];
+			if (s->address > address || address >= s->address + s->size)
+				continue;
+			if ((flags & DRGN_FIND_SYMBOL_NAME) &&
+			    strcmp(s->name, name) != 0)
+				continue;
+			if (!add_symbol_result(builder, s))
+				return &drgn_enomem;
+			if (flags & DRGN_FIND_SYMBOL_ONE)
+				break;
+		}
+	} else if (flags & DRGN_FIND_SYMBOL_NAME) {
+		struct drgn_symbol_name_table_iterator it =
+			drgn_symbol_name_table_search(&index->htab, &name);
+		if (!it.entry)
+			return NULL;
+		for (uint32_t i = it.entry->value.start; i < it.entry->value.end; i++) {
+			struct drgn_symbol *s = &index->symbols[index->name_sort[i]];
+			if (!add_symbol_result(builder, s))
+				return &drgn_enomem;
+			if (flags & DRGN_FIND_SYMBOL_ONE)
+				break;
+		}
+	} else {
+		for (int i = 0; i < index->num_syms; i++) {
+			struct drgn_symbol *s = &index->symbols[i];
+			if (!add_symbol_result(builder, s))
+				return &drgn_enomem;
+			if (flags & DRGN_FIND_SYMBOL_ONE)
+				break;
+		}
+	}
+	return NULL;
+}
+
+void
+drgn_symbol_index_builder_init(struct drgn_symbol_index_builder *builder)
+{
+	builder->names = (struct string_builder)STRING_BUILDER_INIT;
+	symbol_vector_init(&builder->symbols);
+}
+
+void
+drgn_symbol_index_builder_deinit(struct drgn_symbol_index_builder *builder)
+{
+	string_builder_deinit(&builder->names);
+	symbol_vector_deinit(&builder->symbols);
+}
+
+bool
+drgn_symbol_index_builder_add(struct drgn_symbol_index_builder *builder,
+			      const struct drgn_symbol *ptr)
+{
+	struct drgn_symbol copy = *ptr;
+
+	// Temporarily store the index into the name
+	copy.name = (char *)builder->names.len;
+	return string_builder_append(&builder->names, ptr->name)
+		&& string_builder_appendc(&builder->names, '\0')
+		&& symbol_vector_append(&builder->symbols, &copy);
+}
+
+struct drgn_error *
+drgn_symbol_index_init_from_builder(struct drgn_symbol_index *index,
+				    struct drgn_symbol_index_builder *builder)
+{
+	size_t names_len = builder->names.len;
+	char *names = string_builder_steal(&builder->names);
+	names = realloc(names, names_len);
+	if (!names) {
+		symbol_vector_deinit(&builder->symbols);
+		return &drgn_enomem;
+	}
```

Comment:
This leaks the old `names` pointer in the failure case. We can tolerate `realloc` failing, so maybe:
```suggestion
	char *tmp_names = realloc(names, names_len);
	if (tmp_names)
		names = tmp_names;
```